api/v1beta2/awsmachine_types.go

```diff
@@ -197,6 +208,26 @@ type Ignition struct {
 	// +kubebuilder:default="2.3"
 	// +kubebuilder:validation:Enum="2.3";"3.0";"3.1";"3.2";"3.3";"3.4"
 	Version string `json:"version,omitempty"`
+
+	// StorageType defines how to store the boostrap user data for Ignition.
+	// This can be used to instruct Ignition from where to fetch the user data to bootstrap an instance.
+	//
+	// When omitted, the storage option will default to S3.
+	//
+	// When set to "S3", if the capability is available and S3 configuration
+	// is correctly provided in the Cluster object,
+	// an S3 bucket will be used to store bootstrap user data.
+	//
+	// When set to "InstanceUserData", EC2 Instance User Data will be used to store the machine bootstrap user data.
+	// This option is considered less secure than others as user data may contain sensitive informations (keys, certificates, etc.)
+	// and users with ec2:DescribeInstances permission or users running pods
+	// that can access the ec2 metadata service have access to this sensitive information.
+	// So this is only to be used at ones own risk, and only when other more secure options are not viable.
+	//
+	// +optional
+	// +kubebuilder:default="S3"
+	// +kubebuilder:validation:Enum:="S3";"InstanceUserData"
+	StorageType IgnitionStorageTypeOption `json:"storageType,omitempty"`
```

Comment:
```suggestion
	// +optional
	// +kubebuilder:default="ClusterObjectStore"
	// +kubebuilder:validation:Enum:="ClusterObjectStore";"UnencryptedUserData"
	StorageType IgnitionStorageTypeOption `json:"storageType,omitempty"`
```

Proposing that we do away, at this point, from choosing `S3` as an option here because the object storage is defined on the AWSCluster object.

I'd also propose to rename `InstanceUserData` to `UnencryptedUserData` instead, given that the current implementation can be expanded in the future if encryption options are available for Ignition, and `UserData` is a well-known concept associated with an EC2 Instance already.

@damdo @dlipovetsky @richardcase thoughts?

Reply:
Hey @vincepri , thanks for reviewing this.
As we discussed offline, I'm happy with changing the enum value `S3` to `ClusterObjectStore`, as that can give more insight into the fact that the Object store details are defined at a Cluster level on the Cluster object.

As for the `UnencryptedUserData` vs `InstanceUserData`, I don't have strong opinions, can go either way. So I'm happy to change provided others are too.

Let's see if @dlipovetsky and @richardcase agree.

Reply:
I think we need to think consider whether changing the enum here is considered backwards compatible. Especially as this is now potentially suggesting:

- Adding a new enum value (i.e. `UnencryptedUserData`)
- Changing an existing one. Is `ClusterObjectStore` schematically equivalent to `S3`?
- Changing the default value

Let's review the [Changing the API](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md) doc. If we don't have a compatible change, then we'd need to bump the API version.

Its been a whilst since i read the doc, so i will do that ASAP.

Reply:
@richardcase API wise, the `storageType` Enum is a new field that I'm adding under the existing `ignition` field, so in that regard it is an addition (no breaking change).

Behaviour wise, up until now, there was no way to change the default behaviour, which was to use the Cluster Object Storage reference defined under the `AWSCluster` object spec [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/f8b9cfd43ff0dd3e2887c3acad3d1ac53284cf8c/api/v1beta2/awscluster_types.go#L98-L103).
From now on, with this new Enum we will provide a way to tweak this.
If this is not set (default) or `ClusterObjectStore` is set, the user will get the current behaviour.
If `UnencryptedUserData` is set, the user will get a different behaviour (storing of the value unencrypted directly in the AWS User Data).

So no breaking change here both from an API and from a behaviour standpoints, default remains the same, only an extra option for different behaviour.

Reply:
Great, thanks for explaining that @damdo 👍 I will take time to properly review asap.

Reply:
Thanks @richardcase
I haven't yet included the above Enum values renaming that @vincepri is proposing, but I intend to, if we are all happy with it.
So once you let me know, if you are happy, I can change this as well before the final review.

Reply:
After chatting with @nrb @vincepri @richardcase and @dlipovetsky today at the CAPA meeting we decided to go with the Enum values suggested by @vincepri. I'll update the PR tomorrow with those values.